app/uk/gov/hmrc/gform/sharedmodel/formtemplate/Validator.scala

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2017 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.gform.sharedmodel.formtemplate
+
+import cats.data.Validated
+import cats.data.Validated.{ Invalid, Valid }
+import play.api.libs.json._
+import play.api.libs.json.Reads._
+import uk.gov.hmrc.gform.core.parsers.ExprParsers
+import uk.gov.hmrc.play.http.HeaderCarrier
+import play.api.libs.functional.syntax._
+import play.api.libs.functional.syntax._
+
+import scala.concurrent.{ ExecutionContext, Future }
+
+sealed trait Validator {
+  def errorMessage: String
+  def validate(data: Map[FieldId, Seq[String]])(f: ((String, String)) => Future[Boolean])(implicit ec: ExecutionContext, hc: HeaderCarrier): Future[Validated[Map[FieldId, Set[String]], Unit]]
+}
+
+case object Validator {
+
+  val reads: Reads[Validator] = Reads { json =>
+    (json \ "validatorName").as[String] match {
+      case "hmrcUTRPostcodeCheck" => json.validate[HMRCUTRPostcodeCheckValidator]
+    }
+  }
+
+  val writes: OWrites[Validator] = OWrites {
+    case v: HMRCUTRPostcodeCheckValidator => HMRCUTRPostcodeCheckValidator.format.writes(v)
+
+  }
+
+  implicit val format = OFormat(reads, writes)
+
+}
+
+case class HMRCUTRPostcodeCheckValidator(errorMessage: String, utr: FormCtx, postcode: FormCtx) extends Validator {
+
+  val utrFieldId = FieldId(utr.value)
+  val postcodeFieldId = FieldId(postcode.value)
+
+  def validate(data: Map[FieldId, Seq[String]])(f: ((String, String)) => Future[Boolean])(implicit ec: ExecutionContext, hc: HeaderCarrier): Future[Validated[Map[FieldId, Set[String]], Unit]] = {
+    val dataGetter: FieldId => String = id => data.get(id).toList.flatten.headOption.getOrElse("")
```

Comment:
hmm, if there is no data for given id maybe we should immediately return invalid?

Reply:
thats a confusing one, as the postcode can be blank ... but cannot be optional I was speaking to Nathan about this and its confusing.